File: day8/main.py
from __future__ import annotations

import sys
import time
from math import sqrt

from dataclasses import dataclass


@dataclass
class Point:
    x: int
    y: int
    z: int

    def distance(self, other: Point) -> float:
        return sqrt(
            (self.x - other.x) ** 2 + (self.y - other.y) ** 2 + (self.z - other.z) ** 2
        )


@dataclass
class Edge:
    a: int
    b: int
    weight: float


@dataclass
class UnionFind:
    """Union-Find Disjoint Set data structure."""

    def __init__(self, size: int):
        self.parent = [i for i in range(size)]
        self.rank = [0 for _ in range(size)]
        self.set_size = [1 for _ in range(size)]
        self.num_sets = size

    def find_set(self, i: int) -> int:
        if self.parent[i] == i:
            return i
        self.parent[i] = self.find_set(self.parent[i])
        return self.parent[i]

    def is_same_set(self, i: int, j: int) -> bool:
        return self.find_set(i) == self.find_set(j)

    def union_set(self, i: int, j: int) -> None:
        if not self.is_same_set(i, j):
            x = self.find_set(i)
            y = self.find_set(j)

            if self.rank[x] > self.rank[y]:
                self.parent[y] = x
                self.set_size[x] += self.set_size[y]
            else:
                self.parent[x] = y
                self.set_size[y] += self.set_size[x]
                if self.rank[x] == self.rank[y]:
                    self.rank[y] += 1

            self.num_sets -= 1

    def size_of_set(self, i: int) -> int:
        return self.set_size[self.find_set(i)]

    def num_of_sets(self) -> int:
        return self.num_sets


# Parse input
points = []  # type: list[Point]
for line in sys.stdin:
    line = line.strip()
    points.append(Point(*map(int, line.split(","))))

# Construct graph data structure using an edge list representation
edge_list = []  # type: list[Edge]
for i in range(len(points)):
    for j in range(i + 1, len(points)):
        edge_list.append(Edge(i, j, points[i].distance(points[j])))

# Kruskal's algorithm for Minimum Spanning Tree

# Sort edges by increasing weight (distance)
edge_list.sort(key=lambda e: e.weight)
union_find = UnionFind(len(points))

# Join the first 1000 closest points
connections = 1000
for i in range(connections):
    union_find.union_set(edge_list[i].a, edge_list[i].b)

# Extract all sets sizes and find the answer for part 1
set_sizes = [0 for _ in range(len(points))]
for i in range(len(points)):
    set_sizes[union_find.find_set(i)] = union_find.size_of_set(i)
set_sizes.sort(reverse=True)
print(set_sizes[0] * set_sizes[1] * set_sizes[2])

# Continue until there is only one set for part 2
i = connections
while union_find.num_of_sets() > 1:
    union_find.union_set(edge_list[i].a, edge_list[i].b)
    i += 1
i -= 1
print(points[edge_list[i].a].x * points[edge_list[i].b].x)

File: day8/Main.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        ArrayList<Point> points = new ArrayList<>();
        while (scanner.hasNextLine()) {
            String[] comp = scanner.nextLine().split(",");
            points.add(new Point(Integer.parseInt(comp[0]), Integer.parseInt(comp[1]), Integer.parseInt(comp[2])));
        }

        ArrayList<Edge> edges = new ArrayList<>(points.size() * points.size());
        for (int i = 0; i < points.size(); i++) {
            for (int j = i + 1; j < points.size(); j++) {
                edges.add(new Edge(i, j, points.get(i).distance(points.get(j))));
            }
        }

        Collections.sort(edges);

        UnionFind unionFind = new UnionFind(points.size());

        int conns = 1000;
        for (int i = 0; i < conns; i++) {
            unionFind.union(edges.get(i).i, edges.get(i).j);
        }

        int[] setSizes = new int[points.size()];
        for (int i = 0; i < setSizes.length; i++) {
            setSizes[unionFind.find(i)] = unionFind.getSetSize(i);
        }
        Arrays.sort(setSizes);
        System.out.println(setSizes[setSizes.length - 1] * setSizes[setSizes.length - 2] * setSizes[setSizes.length - 3]);

        // Continue until fully connected
        int i;
        for (i = conns; unionFind.getNumSets() > 1; i++) {
            unionFind.union(edges.get(i).i, edges.get(i).j);
        }
        Edge edge = edges.get(i - 1);
        System.out.println(points.get(edge.i).x * points.get(edge.j).x);
    }
}


class Edge implements Comparable<Edge> {
    public int i;
    public int j;
    public int weight;

    public Edge(int i, int j, int weight) {
        this.i = i;
        this.j = j;
        this.weight = weight;
    }

    @Override
    public int compareTo(Edge o) {
        return this.weight - o.weight;
    }
}


class Point {
    public int x;
    public int y;
    public int z;

    public Point(int x, int y, int z) {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    public int distance(Point p) {
        return (int) (Math.pow(this.x - p.x, 2) + Math.pow(this.y - p.y, 2) + Math.pow(this.z - p.z, 2));
    }
}


class UnionFind {
    private final int[] parent;
    private final int[] rank;
    private final int[] setSize;
    private int numSets;

    public UnionFind(int size) {
        parent = new int[size];
        rank = new int[size];
        setSize = new int[size];
        for (int i = 0; i < size; i++) {
            parent[i] = i;
            rank[i] = 0;
            setSize[i] = 1;
        }
        numSets = size;
    }

    public int find(int i) {
        if (parent[i] != i) {
            parent[i] = find(parent[i]);  // Path compression
        }
        return parent[i];
    }

    public void union(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);

        if (rootX == rootY) {
            return;
        }

        if (rank[rootX] > rank[rootY]) {
            parent[rootY] = rootX;
            setSize[rootX] += setSize[rootY];
        } else {
            parent[rootX] = rootY;
            setSize[rootY] += setSize[rootX];
            if (rank[rootX] == rank[rootY]) {
                rank[rootY] += 1;
            }
        }

        numSets--;
    }

    public int getSetSize(int i) {
        return setSize[i];
    }

    public int getNumSets() {
        return numSets;
    }
}
